### PR TITLE
Gutenboarding: Preview fonts

### LINKED ISSUE #40010: Gutenboarding: style picker - font pairing list
Create a list of font pairings with custom font applied to each button:

image

Clicking on font pairing changes font styles in the preview.

**Questions:**
- Should buttons be static images to avoid loading so many fonts?
- Can fonts be loaded so that only the characters we need are loaded?
- Localization concerns?

We have "global styles" plugin wpcom and another one being worked in core Gutenberg. Initially, we'll leverage the wpcom version and we should make sure to be aligned with the core version to make the transition easier some day.

Context p7rd6c-2dV-p2

### PR DESCRIPTION
**Pull request by sirreal**
#### Changes proposed in this Pull Request

Load rendered preview with fonts and dynamically update fonts.

demo

Closes #40010.

#### Testing instructions

Visit `/gutenboarding` in development and play with design previews and font selectors. Only **Rockfield** design is fully implemented.

### CODE REVIEW

File: client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx [hunk 1]
```diff
@@ -2,35 +2,60 @@
...
+	const [ previewHtml, setPreviewHtml ] = React.useState< string >();
+	const iframe = React.useRef< HTMLIFrameElement >( null );
+
+	React.useEffect( () => {
+		const eff = async () => {
+			const [ { fontFamily: headings }, { fontFamily: base } ] = fonts;
+			const url = `https://public-api.wordpress.com/rest/v1/template-demo/${ encodeURIComponent(
+				'rockfield'
+			) }/${ encodeURIComponent( 'rockfield' ) }/en/${ encodeURIComponent(
+				'gym'
```
Comment (simison) [inline client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx]: For a demo `default` also works according to Javi.

File: client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx [hunk 2]
```diff
@@ -11,13 +12,110 @@ import { STORE_KEY } from '../../stores/onboard';
...
 }
 const Preview: React.FunctionComponent< Props > = ( { fonts, viewport } ) => {
-	const { selectedDesign } = useSelect( select => select( STORE_KEY ).getState() );
-	const [ { title: fontA }, { title: fontB } ] = fonts;
+	const [ previewHtml, setPreviewHtml ] = React.useState< string >();
+	const [ requestedFonts, setRequestedFonts ] = React.useState< Set< string > >( new Set() );
+
+	const { selectedDesign, siteVertical } = useSelect( select =>
+		select( STORE_KEY ).getState()
+	) as { selectedDesign: Design; siteVertical: SiteVertical };
```
Comment (ockham) [inline client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx]: Why do we need the cast? :thinking:

File: client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx [hunk 3]
```diff
@@ -11,13 +12,110 @@ import { STORE_KEY } from '../../stores/onboard';
...
+
+	React.useEffect(
+		() => {
+			const eff = async () => {
+				const [ { fontFamily: fontHeadings }, { fontFamily: fontBase } ] = fonts;
+				const templateUrl = `https://public-api.wordpress.com/rest/v1/template/demo/${ encodeURIComponent(
+					selectedDesign.slug
+				) }/${ encodeURIComponent( selectedDesign.slug ) }/`;
+				const url = addQueryArgs( templateUrl, {
+					language: 'en',
```
Comment (ockham) [inline client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx]: Can we `useLangRouteParam` already? (Will the REST API endpoint fall back to `en`?)

File: client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx [hunk 4]
```diff
@@ -11,13 +12,110 @@ import { STORE_KEY } from '../../stores/onboard';
...
+					font_headings: fontHeadings,
+					font_base: fontBase,
+				} );
+				let resp;
+				try {
+					resp = await window.fetch( url );
+					if ( resp.status < 200 || resp.status >= 300 ) {
+						throw resp;
+					}
+				} catch ( err ) {
```
Comment (ockham) [inline client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx]: Note to our future selves, maybe we should start using [React Error Boundaries](https://reactjs.org/docs/error-boundaries.html) in Gutenboarding (for better error handling patterns?)

File: client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx [hunk 5]
```diff
@@ -11,13 +12,110 @@ import { STORE_KEY } from '../../stores/onboard';
...
+				query.push( `family=${ encodeURI( base ) }:${ axis }` );
+				nextFonts.add( base );
+			}
+
+			if ( query.length ) {
+				const l = iframeDocument.createElement( 'link' );
+				l.rel = 'stylesheet';
+				l.type = 'text/css';
+				l.href = `${ baseURL }?${ query.join( '&' ) }&display=swap`;
+				iframeDocument.head.appendChild( l );
```
Comment (ockham) [inline client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx]: Noting here that it'd be nice to encapsulate Google Fonts API handling. There seem to be a few libraries out there to do that ([#](https://www.npmjs.com/package/get-google-fonts#Example-Result)), but I haven't found one that's used widely enough to make sense to use.

There's one [Webpack plugin](https://www.npmjs.com/package/google-fonts-plugin) that seems relatively popular, but I'm not sure we want to add more chunks to the mix.

Comment (simison) [reply client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx]: These bits might/will change once core Global styles is ready; now it's just stying to stay in sync with what happens in wpcom-Global styles and it's ok to be more manual/disconnected.

There might be some core functions available later on for all this.

File: client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx (hunk 4 above)
Comment (sirreal) [reply client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx]: I believe blocks all have error boundaries, but definitely good to keep in mind 👍

File: client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx (hunk 2 above)
Comment (sirreal) [reply client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx]: I had a comment at some point, seems a good practice to accompany cases with comments. I'll add another.

The type is possibly undefined, but our flow should guarantee that on this step we have both of these defined.

File: client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx [hunk 6]
```diff
@@ -2,22 +2,123 @@
...
+
+	React.useEffect(
+		() => {
+			const eff = async () => {
+				const [ { fontFamily: fontHeadings }, { fontFamily: fontBase } ] = fonts;
+				const templateUrl = `https://public-api.wordpress.com/rest/v1/template/demo/${ encodeURIComponent(
+					selectedDesign.slug
+				) }/${ encodeURIComponent( selectedDesign.slug ) }/`;
+				const url = addQueryArgs( templateUrl, {
+					language: language,
```
Comment (ockham) [inline client/landing/gutenboarding/onboarding-block/style-preview/preview.tsx]: Nit

```suggestion
					language,
```

**Comment by sirreal** [comment]
Rebased.

**Pull request merged**